shell/platform/linux/fl_binary_messenger.cc

```diff
@@ -233,7 +220,8 @@ static gboolean send_response(FlBinaryMessenger* messenger,
   g_return_val_if_fail(response_handle->messenger == self, FALSE);
   g_return_val_if_fail(response_handle->response_handle != nullptr, FALSE);
 
-  if (self->engine == nullptr) {
+  g_autoptr(FlEngine) engine = FL_ENGINE(g_weak_ref_get(&self->engine));
```

Comment:
This worries me because I believe this can result in the FlEngine's dispose method to be called on a background thread.  I haven't found clear documentation about what thread calls that.  Potentially we could add a ref here and post a task on the platform thread to unref to guarantee that the dispose happens on the platform thread?

Reply:
I think you're right. Looking at the glib source the dispose will be called by the last thread that called `g_object_unref`.

Reply:
You should be able to use a single reference here, and just ensure that the unref is always done on the main loop.

i.e.
```c
gboolean
do_unref (gpointer value)
{
  g_object_unref(value);
  return G_SOURCE_REMOVE;
}

{
  FlEngine *engine = FL_ENGINE(g_weak_ref_get(&self->engine));

  ...

  g_idle_add(do_unref, engine);
}
```

This looses the autofree behaviour, but potentially we could make a `g_autoptr_unref_in_idle` macro that makes this tidier.